pkg/meta/distributed_lock.go

```diff
@@ -0,0 +1,314 @@
+		return err
+	}
+	if !intest.InTest {
+		var renewCtx context.Context
+		renewCtx, d.cancelRenewLoop = context.WithCancel(d.outerCtx)
+		go d.renewLockLoop(renewCtx)
+	}
+	return nil
+}
+
+func (d *DistributedLock) lockExpired(lc *lockContent, currentStartTS uint64) bool {
+	if intest.InTest {
+		return false
+	}
+	lockStartTime := model.TSConvert2Time(lc.StartTS)
+	currentTime := model.TSConvert2Time(currentStartTS)
+	expired := currentTime.Sub(lockStartTime) > d.lease
+	if expired {
+		logutil.Logger(d.outerCtx).Warn("lock is expired",
+			zap.Time("lockStartTime", lockStartTime),
+			zap.Time("currentTime", currentTime),
+			zap.Duration("lease", d.lease))
+	}
+	return expired
+}
+
+// Unlock unlocks the distributed lock.
+func (d *DistributedLock) Unlock() error {
+	defer func() {
+		if d.cancelRenewLoop != nil {
+			d.cancelRenewLoop()
+		}
+	}()
+	return d.getLockInTxnWithRetry(context.Background(),
+		func(lc *lockContent, txn DataTxn) (storeErr, otherErr error) {
+			if lc == nil || lc.InstanceID != d.instanceID {
+				logutil.Logger(d.outerCtx).Warn("lock is unexpectedly removed", zap.String("instanceID", d.instanceID))
+				return nil, nil
+			}
+
+			// Lock is exist, remove it.
+			err := txn.Del(d.lockName)
+			if err != nil {
+				return errors.Trace(err), nil
+			}
+			return nil, nil
+		})
+}
+
+func (d *DistributedLock) getLockInTxnWithRetry(ctx context.Context, fn lockHandleFunc) error {
+	retryCnt := 0
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-d.outerCtx.Done():
+			// We only check the outer context error after retry 10 times,
+			// so that the lock can be released even if the outer context is canceled.
+			if retryCnt >= 10 {
+				return d.outerCtx.Err()
+			}
+		default:
+		}
+
+		var noRetry bool
+		err := runInTxn(ctx, d.store, func(txn DataTxn) error {
+			data, err := txn.Get(d.lockName)
+			if err != nil {
+				return errors.Trace(err)
+			}
+
+			lc, err := decodeLockContent(data)
+			if err != nil {
+				noRetry = true
+				return errors.Trace(err)
+			}
+
+			storeErr, otherErr := fn(lc, txn)
+			if otherErr != nil {
+				noRetry = true
+				return errors.Trace(otherErr)
+			}
+			return storeErr
+		})
+
+		if err != nil {
+			if err == lockHeldErr {
+				time.Sleep(d.backoff)
+				continue
+			}
+			logutil.Logger(d.outerCtx).Info("lock encounter error",
+				zap.Error(err), zap.Bool("retry", !noRetry))
+			if noRetry {
+				return err
+			}
+			retryCnt++
+			time.Sleep(d.backoff)
+			continue
+		}
+		return nil
+	}
+}
+
+type lockHandleFunc func(lc *lockContent, txn DataTxn) (storeErr, otherErr error)
+type txnHandleFunc func(txn DataTxn) error
+
+func runInTxn(ctx context.Context, store DataStore, fn txnHandleFunc) error {
+	txn, err := store.Begin(ctx)
+	if err != nil {
+		return err
+	}
+	err = fn(txn)
+	if err != nil {
+		txn.Rollback()
+		return err
+	}
+	return txn.Commit()
+}
+
+func decodeLockContent(data []byte) (*lockContent, error) {
+	if len(data) == 0 {
```

Comment:
in which case len(data) is 0 and txn.Get has no error?

Reply:
That is why I change the implementation of `DataTxn.Get()`:

```
if kv.ErrNotExist.Equal(err) {
	err = nil
}
```